### ISSUE TITLE
Class loading may fail when JVM is stopped using SIGINT

### ISSUE THREAD

**Issue by wilkinsona** [spring-projects/spring-boot#38154]
When the JVM is stopped using SIGINT, Boot's shutdown hook runs and closes the application context. Class loading that occurs during this close processing may fail due to a thread being interrupted while calling `FileChannel.read(ByteBuffer, long)`. The interruption causes the read to close the channel and fail with a `ClosedByInterruptException`. This leaves `FileChannelDataBlock` with a `FileChannel` that should be open according to its reference count but isn't.

**Comment by wilkinsona** [comment]
Possible fix: https://github.com/wilkinsona/spring-boot/tree/gh-38154

**Comment by philwebb** [comment]
The `tracker` stuff is only used in tests to I think we could get away without tracking the reopen if want to.

**Comment by wilkinsona** [comment]
Thanks for having a look, Phil. I think I might be confused looking at the code in the future if we omit the tracking so I'd prefer to keep it. I've pushed a commit that refines the commit as I think it was a mistake to reset the thread's interrupted flag. Instead, we now just repair the file channel and re-throw the exception. This allows the interruption on the current thread to continue will also ensuring that the file channel is usable on other threads. I've tested it with @onobc's sample that brought the problem to my attention and it seems to work.
